cloudinit/distros/alpine.py

```diff
@@ -202,6 +204,423 @@ def preferred_ntp_clients(self):
 
         return self._preferred_ntp_clients
 
+    def add_user(self, name, **kwargs):
+        """
+        Add a user to the system using standard tools
+
+        On Alpine this may use either useradd or adduser depending on whether
+        the shadow package is installed.
+        """
+        if util.is_user(name):
+            LOG.info("User %s already exists, skipping.", name)
+            return
+
+        if "create_groups" in kwargs:
+            create_groups = kwargs.pop("create_groups")
+        else:
+            create_groups = True
+
+        # Decide whether 'useradd' or 'adduser' will be used
+        if subp.which("useradd"):
```

Comment:
If we're using `useradd`, does the default implementation work? Instead of separate paths for both, at the top of this method could we put something like:
```python
if subp.which("useradd"):
    return super().add_user(name, **kwargs)
```

Reply:
Hmm. I hadn't considered that.

This Alpine-specific function doesn't define "--selinux-user" in the list of valid useradd options for example as Alpine doesn't support Selinux.

Also a lot of the code in the Alpine-specific add_user function would still need to remain for "adduser" support so I'm not convinced of the benefits of splitting "useradd" and "adduser" between 2 separate functions.

Reply:
When I walked through it, almost all of the `useradd` code was essentially just duplicated into the alpine code behind if statements. For `--selinux-user`, we can pop that off before calling it. You sure something like this wouldn't work?
```diff
diff --git a/cloudinit/distros/alpine.py b/cloudinit/distros/alpine.py
index 1b8d52c27..9d88a4b30 100644
--- a/cloudinit/distros/alpine.py
+++ b/cloudinit/distros/alpine.py
@@ -11,7 +11,7 @@ import os
 import re
 import stat
 from datetime import datetime
-from typing import Optional
+from typing import Any, Dict, Optional
 
 from cloudinit import distros, helpers, subp, util
 from cloudinit.distros.parsers.hostname import HostnameConf
@@ -215,64 +215,36 @@ class Distro(distros.Distro):
             LOG.info("User %s already exists, skipping.", name)
             return
 
+        if "selinux_user" in kwargs:
+            LOG.warning("Ignoring selinux_user parameter for Alpine Linux")
+            del kwargs["selinux_user"]
+
+        if subp.which("useradd"):
+            return super().add_user(name, **kwargs)
+
         if "create_groups" in kwargs:
             create_groups = kwargs.pop("create_groups")
         else:
             create_groups = True
 
-        # Decide whether 'useradd' or 'adduser' will be used
-        if subp.which("useradd"):
-            # Shadow useradd
-            use_useradd = True
-        else:
-            # Busybox adduser
-            use_useradd = False
-
         # Since we are creating users, we want to carefully validate the
         # inputs. If something goes wrong, we can end up with a system
         # that nobody can login to.
-        if use_useradd:
-            # Shadow package is installed, use 'useradd'
-            add_cmd = ["useradd", name]
-            log_add_cmd = ["useradd", name]
-
-            add_opts = {
-                "expiredate": "--expiredate",
-                "gecos": "--comment",
-                "groups": "--groups",
-                "homedir": "--home",
-                "inactive": "--inactive",
-                "passwd": "--password",
-                "primary_group": "--gid",
-                "shell": "--shell",
-                "uid": "--uid",
-            }
-
-            add_flags = {
-                "no_log_init": "--no-log-init",
-                "no_user_group": "--no-user-group",
-                "system": "--system",
-            }
-        else:
-            # Shadow package is not installed, use Busybox 'adduser'
-            add_cmd = ["adduser", "-D"]
-
-            add_opts = {
-                "expiredate": "",
-                "gecos": "-g",
-                "groups": "",
-                "homedir": "-h",
-                "inactive": "",
-                "passwd": "",
-                "primary_group": "-G",
-                "shell": "-s",
-                "uid": "-u",
-            }
-
-            add_flags = {"system": "-S"}
-
-        if use_useradd:
-            redact_opts = ["passwd"]
+        add_cmd = ["adduser", "-D"]
+
+        add_opts = {
+            "expiredate": "",
+            "gecos": "-g",
+            "groups": "",
+            "homedir": "-h",
+            "inactive": "",
+            "passwd": "",
+            "primary_group": "-G",
+            "shell": "-s",
+            "uid": "-u",
+        }
+
+        add_flags = {"system": "-S"}
 
         # support kwargs having groups=[list] or groups="g1,g2"
         groups = kwargs.get("groups")
@@ -309,148 +281,114 @@ class Distro(distros.Distro):
         if "uid" in kwargs.keys():
             kwargs["uid"] = str(kwargs["uid"])
 
-        addn_groups = None
-        expirydate = None
-        inactive = None
-        passwd = None
+        unsupported_busybox_values: Dict[str, Any] = {
+            "groups": None,
+            "expirydate": None,
+            "inactive": None,
+            "passwd": None,
+        }
 
         # Check the values and create the command
         for key, val in sorted(kwargs.items()):
-
             if key in add_opts and val and isinstance(val, str):
-                if not use_useradd and (
-                    key == "expiredate"
-                    or key == "groups"
-                    or key == "inactive"
-                    or key == "passwd"
-                ):
+                if key in unsupported_busybox_values:
                     # Busybox 'adduser' does not support specifying these
                     # options so store them for use via alternative means.
-                    if key == "expiredate":
-                        expirydate = val
-                    elif key == "groups":
-                        addn_groups = val.split(",")
-                    elif key == "inactive":
-                        inactive = val
-                    elif key == "passwd":
-                        passwd = val
+                    if key == "groups":
+                        unsupported_busybox_values[key] = val.split(",")
+                    else:
+                        unsupported_busybox_values[key] = val
                 else:
                     add_cmd.extend([add_opts[key], val])
 
-                if use_useradd:
-                    # Redact certain fields from the logs
-                    if key in redact_opts:
-                        log_add_cmd.extend([add_opts[key], "REDACTED"])
-                    else:
-                        log_add_cmd.extend([add_opts[key], val])
-
             elif key in add_flags and val:
                 add_cmd.append(add_flags[key])
-                if use_useradd:
-                    log_add_cmd.append(add_flags[key])
 
         # Don't create the home directory if directed so or if the user is a
         # system user
         if kwargs.get("no_create_home") or kwargs.get("system"):
-            if use_useradd:
-                # Shadow useradd
-                add_cmd.append("-M")
-                log_add_cmd.append("-M")
-            else:
-                # Busybox adduser
-                add_cmd.append("-H")
-        else:
-            if use_useradd:
-                # Shadow useradd
-                add_cmd.append("-m")
-                log_add_cmd.append("-m")
+            # Busybox adduser
+            add_cmd.append("-H")
 
-        if not use_useradd:
-            # Busybox 'adduser' puts username at end of command
-            add_cmd.append(name)
+        # Busybox 'adduser' puts username at end of command
+        add_cmd.append(name)
 
         # Run the command
-        LOG.debug("Adding user %s", name)
-        if use_useradd:
-            try:
-                subp.subp(add_cmd, logstring=log_add_cmd)
-            except Exception as e:
-                util.logexc(LOG, "Failed to create user %s", name)
-                raise e
-        else:
-            try:
-                subp.subp(add_cmd)
-            except Exception as e:
-                util.logexc(LOG, "Failed to create user %s", name)
-                raise e
+        try:
+            subp.subp(add_cmd)
+        except Exception as e:
+            util.logexc(LOG, "Failed to create user %s", name)
+            raise e
 
-        if not use_useradd:
-            if addn_groups is not None:
-                # Separately add user to each additional group as Busybox
-                #  'adduser' does not support specifying additional groups.
-                for addn_group in addn_groups:
-                    LOG.debug("Adding user to group %s", addn_group)
-                    try:
-                        subp.subp(["addgroup", name, addn_group])
-                    except Exception as e:
-                        util.logexc(
-                            LOG,
-                            "Failed to add user %s to group %s",
-                            name,
-                            addn_group,
-                        )
-                        raise e
-
-            if passwd is not None:
-                # Separately set password as Busybox 'adduser' does not
-                # support passing password as CLI option.
-                self.set_passwd(name, passwd, hashed=True)
-
-            if expirydate is not None or inactive is not None:
-                # Separately set expirydate and/or inactive values directly
-                # in /etc/shadow file as Busybox 'adduser' does not support
-                # passing these as CLI options.
-                shadow_contents = None
-                with open("/etc/shadow", "r") as file:
-                    shadow_contents = file.read()
-
-                # Find the line for the user
-                original_line = None
-                for line in shadow_contents.splitlines():
-                    current_user = line.split(":")[0]
-                    if current_user == name:
-                        original_line = line
-
-                # Modify field(s) in copy of user's shadow file entry
-                new_list = original_line.split(":")
-                update_type = ""
-                if expirydate is not None:
-                    # Convert date into number of days since 1st Jan 1970
-                    days = (
-                        datetime.fromisoformat(expirydate)
-                        - datetime.fromisoformat("1970-01-01")
-                    ).days
-                    new_list[7] = str(days)
-                    update_type = "acct expiration date"
-                if inactive is not None:
-                    new_list[6] = inactive
-                    if update_type != "":
-                        update_type = update_type + " & "
-                    update_type = update_type + "inactivity period"
-
-                new_line = ":".join(new_list)
+        # Process remaining options that Busybox 'adduser' does not support
+        addn_groups = unsupported_busybox_values["groups"]
+        if addn_groups is not None:
+            # Separately add user to each additional group as Busybox
+            #  'adduser' does not support specifying additional groups.
+            for addn_group in addn_groups:
+                LOG.debug("Adding user to group %s", addn_group)
+                try:
+                    subp.subp(["addgroup", name, addn_group])
+                except Exception as e:
+                    util.logexc(
+                        LOG,
+                        "Failed to add user %s to group %s",
+                        name,
+                        addn_group,
+                    )
+                    raise e
+
+        passwd = unsupported_busybox_values["passwd"]
+        if passwd is not None:
+            # Separately set password as Busybox 'adduser' does not
+            # support passing password as CLI option.
+            self.set_passwd(name, passwd, hashed=True)
+
+        expirydate = unsupported_busybox_values["expirydate"]
+        inactive = unsupported_busybox_values["inactive"]
+        if expirydate is not None or inactive is not None:
+            # Separately set expirydate and/or inactive values directly
+            # in /etc/shadow file as Busybox 'adduser' does not support
+            # passing these as CLI options.
+            shadow_contents = None
+            with open("/etc/shadow", "r") as file:
+                shadow_contents = file.read()
 
-                # Replace existing line for user with modified line
-                shadow_contents = shadow_contents.replace(
-                    original_line, new_line
-                )
-                LOG.debug("Updating %s for user %s", update_type, name)
-                util.write_file(
-                    "/etc/shadow",
-                    shadow_contents,
-                    omode="w",
-                    preserve_mode=True,
-                )
+            # Find the line for the user
+            original_line = None
+            for line in shadow_contents.splitlines():
+                current_user = line.split(":")[0]
+                if current_user == name:
+                    original_line = line
+
+            # Modify field(s) in copy of user's shadow file entry
+            new_list = original_line.split(":")
+            update_type = ""
+            if expirydate is not None:
+                # Convert date into number of days since 1st Jan 1970
+                days = (
+                    datetime.fromisoformat(expirydate)
+                    - datetime.fromisoformat("1970-01-01")
+                ).days
+                new_list[7] = str(days)
+                update_type = "acct expiration date"
+            if inactive is not None:
+                new_list[6] = inactive
+                if update_type != "":
+                    update_type = update_type + " & "
+                update_type = update_type + "inactivity period"
+
+            new_line = ":".join(new_list)
+
+            # Replace existing line for user with modified line
+            shadow_contents = shadow_contents.replace(original_line, new_line)
+            LOG.debug("Updating %s for user %s", update_type, name)
+            util.write_file(
+                "/etc/shadow",
+                shadow_contents,
+                omode="w",
+                preserve_mode=True,
+            )
 
     def lock_passwd(self, name):
         """
```

It's a lot simpler to me and worked for me with a quick test.

Reply:
I've rewritten the code to make use of the default functions where possible.